src/domain/relay/limit-addresses.mapper.ts

```diff
@@ -2,66 +2,110 @@ import { Inject, Injectable } from '@nestjs/common';
 import { SafeContractHelper } from '@/domain/relay/contracts/safe-contract.helper';
-import { ILoggingService, LoggingService } from '@/logging/logging.interface';
+import { ISafeRepository } from '@/domain/safe/safe.repository.interface';
 import { MultiSendDecoder } from '@/domain/contracts/contracts/multi-send-decoder.helper';
 import { ProxyFactoryDecoder } from '@/domain/relay/contracts/proxy-factory-decoder.helper';
 import {
   getSafeSingletonDeployment,
   getSafeL2SingletonDeployment,
+  getMultiSendCallOnlyDeployment,
 } from '@safe-global/safe-deployments';
 import { SafeDecoder } from '@/domain/contracts/contracts/safe-decoder.helper';
-import { isHex } from 'viem';
+import { isAddress, isHex } from 'viem';
 
 @Injectable()
 export class LimitAddressesMapper {
   // TODO: Abstract with that from SafeContractHelper
   private static SUPPORTED_SAFE_VERSION = '1.3.0';
 
   constructor(
-    @Inject(LoggingService) private readonly loggingService: ILoggingService,
+    @Inject(ISafeRepository)
+    private readonly safeRepository: ISafeRepository,
     private readonly safeContract: SafeContractHelper,
     private readonly erc20Contract: Erc20ContractHelper,
     private readonly safeDecoder: SafeDecoder,
     private readonly multiSendDecoder: MultiSendDecoder,
     private readonly proxyFactoryDecoder: ProxyFactoryDecoder,
   ) {}
 
-  getLimitAddresses(relayPayload: {
+  async getLimitAddresses(args: {
     chainId: string;
     to: string;
     data: string;
-  }): readonly Hex[] {
-    if (!isHex(relayPayload.to) || !isHex(relayPayload.data)) {
+  }): Promise<readonly Hex[]> {
+    // Ensure that the recipient and calldata are valid
+    if (!isAddress(args.to) || !isHex(args.data)) {
       // TODO: Add test coverage once https://github.com/safe-global/safe-client-gateway/pull/1144 is merged
       throw Error('Invalid recipient or calldata provided');
     }
 
-    if (this.isValidExecTransactionCall(relayPayload.to, relayPayload.data)) {
-      return [relayPayload.to];
+    // Calldata matches that of execTransaction and meets validity requirements
+    if (
+      this.isValidExecTransactionCall({
+        to: args.to,
+        data: args.data,
+      })
+    ) {
+      // Safe attempting to relay is official
+      const isOfficial = await this.isOfficialMastercopy({
+        chainId: args.chainId,
+        address: args.to,
+      });
+
+      if (!isOfficial) {
+        throw Error('Invalid Safe contract');
+      }
+
+      // Safe targetted by execTransaction will be limited
+      return [args.to];
     }
 
-    if (this.multiSendDecoder.isMultiSend(relayPayload.data)) {
-      // Validity of MultiSend is part of address retrieval
-      const safeAddress = this.getSafeAddressFromMultiSend(relayPayload.data);
+    // Calldata matches that of multiSend and is from an official MultiSend contract
+    if (
+      this.multiSendDecoder.isMultiSend(args.data) &&
+      this.isOfficialMultiSendDeployment({
```

Comment:
May it also make sense to have a specific error message when `isMultiSend` is `true` but `isOfficialMultiSendDeployment` is false?

Reply:
I separated the errors in ca9b11ca52cb5aabda07baca59bb1391fb567ec6.